# Superstore Sales Analysis - Data Cleaning

**Goal:** load the raw Superstore dataset, check its quality, remove problems, and save a clean copy for the EDA, SQL, and Power BI steps.

**Steps:**
1. Load and inspect the raw data
2. Remove duplicate rows and trim whitespace
3. Validate the cleaned data
4. Save the cleaned data as a formatted Excel file and as a CSV file

## 1. Load and Inspect the Raw Data

In [1]:
import pandas as pd

# Load the raw dataset
df = pd.read_csv("../data/Raw/SampleSuperstore.csv")

# Display first 5 rows
print("First 5 rows:")
display(df.head())

# Dataset shape
print("\nDataset Shape:")
print(df.shape)

# Column names
print("\nColumn Names:")
print(df.columns.tolist())

# Dataset information
print("\nDataset Information:")
df.info()

# Missing values
print("\nMissing Values:")
print(df.isnull().sum())

# Duplicate rows
print("\nDuplicate Rows:")
print(df.duplicated().sum())

First 5 rows:


,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164



Dataset Shape:
(9994, 13)

Column Names:
['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount', 'Profit']

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), objec

**Findings:** the raw file has **9,994 rows and 13 columns**. There are **no missing values**, and the data types are correct (text columns, numeric Sales, Quantity, Discount and Profit). The inspection found **17 duplicate rows**, which are removed in the next step.

## 2. Clean the Data

In [2]:
# ============================================
# SUPERSTORE DATA CLEANING
# ============================================

# Create a copy of the raw data
cleaned_df = df.copy()

# Count duplicate rows
duplicate_count = cleaned_df.duplicated().sum()
print(f"Duplicate rows found: {duplicate_count}")

# Remove exact duplicate rows
cleaned_df = cleaned_df.drop_duplicates()

# Check missing values
print("\nMissing values after duplicate removal:")
print(cleaned_df.isnull().sum())

# Remove unnecessary whitespace from text columns
text_columns = cleaned_df.select_dtypes(include="object").columns

for column in text_columns:
    cleaned_df[column] = cleaned_df[column].str.strip()

# Check data types
print("\nData types:")
print(cleaned_df.dtypes)

# Check cleaned dataset shape
print("\nCleaned dataset shape:")
print(cleaned_df.shape)

# Check duplicates again
print("\nDuplicate rows after cleaning:")
print(cleaned_df.duplicated().sum())

Duplicate rows found: 17

Missing values after duplicate removal:
Ship Mode       0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Category        0
Sub-Category    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

Data types:
Ship Mode        object
Segment          object
Country          object
City             object
State            object
Postal Code       int64
Region           object
Category         object
Sub-Category     object
Sales           float64
Quantity          int64
Discount        float64
Profit          float64
dtype: object

Cleaned dataset shape:
(9977, 13)

Duplicate rows after cleaning:
0


**Result:** the 17 exact duplicate rows were removed and extra spaces were trimmed from text columns. The cleaned dataset has **9,977 rows** and still has 0 missing values.

## 3. Validate the Cleaned Data

In [3]:
# ============================================
# VALIDATE CLEANED DATA
# ============================================

print("First 5 rows of cleaned data:")
display(cleaned_df.head())

print("\nLast 5 rows of cleaned data:")
display(cleaned_df.tail())

print("\nCleaned dataset shape:")
print(cleaned_df.shape)

print("\nMissing values:")
print(cleaned_df.isnull().sum().sum())

print("\nDuplicate rows:")
print(cleaned_df.duplicated().sum())

print("\nBasic statistics:")
display(cleaned_df.describe())

First 5 rows of cleaned data:


,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164



Last 5 rows of cleaned data:


,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
9989,Second Class,Consumer,United States,Miami,Florida,33180,South,Furniture,Furnishings,25.248,3,0.2,4.1028
9990,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Furniture,Furnishings,91.960,2,0.0,15.6332
9991,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Technology,Phones,258.576,2,0.2,19.3932
9992,Standard Class,Consumer,United States,Costa Mesa,California,92627,West,Office Supplies,Paper,29.600,4,0.0,13.3200
9993,Second Class,Consumer,United States,Westminster,California,92683,West,Office Supplies,Appliances,243.160,2,0.0,72.9480



Cleaned dataset shape:
(9977, 13)

Missing values:
0

Duplicate rows:
0

Basic statistics:


,Postal Code,Sales,Quantity,Discount,Profit
count,9977.000000,9977.000000,9977.000000,9977.000000,9977.00000
mean,55154.964117,230.148902,3.790719,0.156278,28.69013
std,32058.266816,623.721409,2.226657,0.206455,234.45784
min,1040.000000,0.444000,1.000000,0.000000,-6599.97800
25%,23223.000000,17.300000,2.000000,0.000000,1.72620
50%,55901.000000,54.816000,3.000000,0.200000,8.67100
75%,90008.000000,209.970000,5.000000,0.200000,29.37200
max,99301.000000,22638.480000,14.000000,0.800000,8399.97600


**Validation:** shape is **(9,977, 13)**, there are **0 missing values** and **0 duplicate rows**. The statistics look reasonable (Discount between 0 and 0.80, Quantity between 1 and 14). Profit has large negative and positive values, which are real sales results and are kept.

## 4. Save the Cleaned Data (Excel)

In [4]:
# ============================================
# SAVE CLEANED DATASET AS FORMATTED EXCEL
# ============================================

import os
from openpyxl import load_workbook
from openpyxl.styles import Font, PatternFill, Alignment
from openpyxl.worksheet.table import Table, TableStyleInfo

# Create Cleaned folder if it doesn't exist
cleaned_folder = "../data/Cleaned"
os.makedirs(cleaned_folder, exist_ok=True)

# Excel file path
cleaned_file_path = "../data/Cleaned/Superstore_Cleaned.xlsx"

# Save cleaned DataFrame to Excel
cleaned_df.to_excel(
    cleaned_file_path,
    index=False,
    sheet_name="Cleaned_Data"
)

# Open the Excel file for formatting
wb = load_workbook(cleaned_file_path)
ws = wb["Cleaned_Data"]


# ============================================
# 1. FORMAT HEADER
# ============================================

for cell in ws[1]:
    cell.font = Font(
        bold=True,
        color="FFFFFF"
    )

    cell.fill = PatternFill(
        fill_type="solid",
        fgColor="1F4E78"
    )

    cell.alignment = Alignment(
        horizontal="center",
        vertical="center"
    )


# ============================================
# 2. SET HEADER ROW HEIGHT
# ============================================

ws.row_dimensions[1].height = 25


# ============================================
# 3. ADJUST COLUMN WIDTH AUTOMATICALLY
# ============================================

for column_cells in ws.columns:

    max_length = 0
    column_letter = column_cells[0].column_letter

    for cell in column_cells:
        if cell.value is not None:
            max_length = max(
                max_length,
                len(str(cell.value))
            )

    # Add extra space
    adjusted_width = max_length + 2

    # Prevent extremely wide columns
    adjusted_width = min(adjusted_width, 30)

    ws.column_dimensions[column_letter].width = adjusted_width


# ============================================
# 4. SET ROW HEIGHT
# ============================================

for row in range(2, ws.max_row + 1):
    ws.row_dimensions[row].height = 20


# ============================================
# 5. FORMAT SALES
# ============================================

sales_column = cleaned_df.columns.get_loc("Sales") + 1
sales_letter = ws.cell(
    row=1,
    column=sales_column
).column_letter

for cell in ws[sales_letter][1:]:
    cell.number_format = '#,##0.00'


# ============================================
# 6. FORMAT PROFIT
# ============================================

profit_column = cleaned_df.columns.get_loc("Profit") + 1
profit_letter = ws.cell(
    row=1,
    column=profit_column
).column_letter

for cell in ws[profit_letter][1:]:
    cell.number_format = '#,##0.00'


# ============================================
# 7. FORMAT DISCOUNT AS PERCENTAGE
# ============================================

discount_column = cleaned_df.columns.get_loc("Discount") + 1
discount_letter = ws.cell(
    row=1,
    column=discount_column
).column_letter

for cell in ws[discount_letter][1:]:
    cell.number_format = '0%'


# ============================================
# 8. CENTER NUMERIC COLUMNS
# ============================================

for row in ws.iter_rows(min_row=2):

    # Postal Code
    row[5].alignment = Alignment(horizontal="center")

    # Quantity
    row[10].alignment = Alignment(horizontal="center")

    # Discount
    row[11].alignment = Alignment(horizontal="center")


# ============================================
# 9. FREEZE HEADER ROW
# ============================================

ws.freeze_panes = "A2"


# ============================================
# 10. ADD FILTERABLE EXCEL TABLE
# ============================================

last_row = ws.max_row
last_column = ws.max_column

table_reference = (
    f"A1:{ws.cell(row=last_row, column=last_column).coordinate}"
)

table = Table(
    displayName="SuperstoreCleanedData",
    ref=table_reference
)

table_style = TableStyleInfo(
    name="TableStyleMedium2",
    showFirstColumn=False,
    showLastColumn=False,
    showRowStripes=True,
    showColumnStripes=False
)

table.tableStyleInfo = table_style

ws.add_table(table)


# ============================================
# 11. SAVE FINAL FILE
# ============================================

wb.save(cleaned_file_path)


# ============================================
# 12. CONFIRMATION
# ============================================

print("============================================")
print("CLEANED EXCEL FILE SAVED SUCCESSFULLY")
print("============================================")
print(f"File location: {cleaned_file_path}")
print(f"Rows saved: {len(cleaned_df)}")
print(f"Columns saved: {len(cleaned_df.columns)}")
print(f"Duplicates: {cleaned_df.duplicated().sum()}")
print(f"Missing values: {cleaned_df.isnull().sum().sum()}")

CLEANED EXCEL FILE SAVED SUCCESSFULLY
File location: ../data/Cleaned/Superstore_Cleaned.xlsx
Rows saved: 9977
Columns saved: 13
Duplicates: 0
Missing values: 0


## 5. Save the Cleaned Data (CSV)

The CSV copy is used to import the data into PostgreSQL for the SQL analysis (PostgreSQL cannot read `.xlsx` files directly).

In [5]:
# ============================================
# SAVE CLEANED DATASET AS CSV (FOR POSTGRESQL)
# ============================================

cleaned_csv_path = "../data/Cleaned/Superstore_Cleaned.csv"

cleaned_df.to_csv(cleaned_csv_path, index=False)

print("============================================")
print("CLEANED CSV FILE SAVED SUCCESSFULLY")
print("============================================")
print(f"File location: {cleaned_csv_path}")
print(f"Rows saved: {len(cleaned_df)}")
print(f"Columns saved: {len(cleaned_df.columns)}")

CLEANED CSV FILE SAVED SUCCESSFULLY
File location: ../data/Cleaned/Superstore_Cleaned.csv
Rows saved: 9977
Columns saved: 13


## Summary

| Item | Value |
|---|---:|
| Original rows | 9,994 |
| Duplicate rows removed | 17 |
| Final cleaned rows | 9,977 |
| Columns | 13 |
| Missing values | 0 |

The cleaned data is saved in `data/Cleaned/` as `Superstore_Cleaned.xlsx` (used by the Python notebooks and Power BI) and `Superstore_Cleaned.csv` (used for the PostgreSQL import).